src/libstore/build.cc

```diff
@@ -4408,6 +4411,19 @@ void SubstitutionGoal::tryNext()
         throw;
     }
 
+    if (info->path != storePath) {
+        if (info->isContentAddressed(*sub)) {
+            auto info2 = std::const_pointer_cast<ValidPathInfo>(std::shared_ptr<const ValidPathInfo>(info));
+            info2->path = storePath.clone();
```

Comment:
This looks dangerous. Isn't this changing the `ValidPathInfo` object that might point to the store's `ValidPathInfo` cache? Maybe just make a copy (`auto info2 = std::make_shared<ValidPathInfo>(*info)`).

Reply:
We should check here that the path has no references, since we can't rewrite inter-store references.

Reply:
Yes. Though I do look forward to later recursively rewriting an all-CA closure provided the store path is the same length!